---
title: "Seurat - Standard workflow and Demultiplexing with hashtags oligos (HTOs) mesonephros"
output:
  html_document:
    keep_md: true
    smart: false
    toc: true
    toc_float: true
    theme: united
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
***


### Resources
https://satijalab.org/seurat/articles/hashing_vignette.html

### Libraries
```{r, message=FALSE}
## Libraries
suppressMessages(library(gridExtra))
suppressMessages(library(ggplot2))
suppressMessages(library(Seurat))
suppressMessages(library(ggplot2))
suppressMessages(library(plotly))
suppressMessages(library(dplyr))
# suppressMessages(library(genefilter))
suppressMessages(library(gplots))

# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.
# We can segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

outputDir  = getwd()
```

## Load sample_b
* Sample b has 3 male mesonephros samples
* Create the Seurat object
* Add HTO data as an idependent assay
* To avoid error message from HTODemux, remove hashtag 6 that has extremely low counts
`Error in HTODemux(object, assay = "HTO", positive.quantile = 0.99) : 
  Cells with zero counts exist as a cluster.`

```{r}
data <- Read10X_h5("/path/to/mesonephros/outs/filtered_feature_bc_matrix.h5")
object <- CreateSeuratObject(counts = data$`Gene Expression`)

## Remove hashtag 6, row 1 
hto_counts <- data$`Antibody Capture`
hto_counts <- hto_counts[2:3, ]
object[['HTO']] <- CreateAssayObject(counts = hto_counts)
object <- NormalizeData(object, assay = "HTO", normalization.method = "CLR")
object
```

## QC
* Remove cells expressing less than 500 genes. We do not set an upper limit for now, as we have seen germ cells expressing more genes than soma cells and we would like to keep both in
* Remove cells with more than 10% MT
* Calculate the percentage of cell cycle genes (same function as for MT genes)
```{r}
object[["percent.mt"]] <- PercentageFeatureSet(object, 
                                               pattern = "^MT-")
cc_genes_in_dataset <- c(s.genes, g2m.genes)[c(s.genes, g2m.genes) %in% rownames(object)]
object[["percent.cc_genes"]] <- PercentageFeatureSet(object, 
                                                     features = cc_genes_in_dataset)
VlnPlot(object, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), # "percent.cc_genes"), 
        ncol = 3, pt.size = 0.1) 

FeatureScatter(object, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(object, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

```{r}
object <- subset(object, subset = nFeature_RNA > 500 & 
    percent.mt < 10)
VlnPlot(object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
    ncol = 3, pt.size = 0.1)
object
```

## Dissociation genes on normalizeds data

* Calculate the expression of dissociation genes as a fraction of total expression (for details see 10.1038/nmeth.4437)
* We calculate it on normalized counts
```{r}
object <- NormalizeData(object, 
                        normalization.method = "LogNormalize", 
                        scale.factor = 50000)
genesChrom <- c("Actg1__chr11", "Ankrd1__chr19", "Arid5a__chr1", "Atf3__chr1", "Atf4__chr15", 
    "Bag3__chr7", "Bhlhe40__chr6", "Brd2__chr17", "Btg1__chr10", "Btg2__chr1", "Ccnl1__chr3", 
    "Ccrn4l__chr3", "Cebpb__chr2", "Cebpd__chr16", "Cebpg__chr7", "Csrnp1__chr9", 
    "Cxcl1__chr5", "Cyr61__chr3", "Dcn__chr10", "Ddx3x__chrX", "Ddx5__chr11", "Des__chr1", 
    "Dnaja1__chr4", "Dnajb1__chr8", "Dnajb4__chr3", "Dusp1__chr17", "Dusp8__chr7", 
    "Egr1__chr18", "Egr2__chr10", "Eif1__chr11", "Eif5__chr12", "Erf__chr7", "Errfi1__chr4", 
    "Fam132b__chr1", "Fos__chr12", "Fosb__chr7", "Fosl2__chr5", "Gadd45a__chr6", 
    "Gcc1__chr6", "Gem__chr4", "H3f3b__chr11", "Hipk3__chr2", "Hsp90aa1__chr12", 
    "Hsp90ab1__chr17", "Hspa1a__chr17", "Hspa1b__chr17", "Hspa5__chr2", "Hspa8__chr9", 
    "Hspb1__chr5", "Hsph1__chr5", "Id3__chr4", "Idi1__chr13", "Ier2__chr8", "Ier3__chr17", 
    "Ifrd1__chr12", "Il6__chr5", "Irf1__chr11", "Irf8__chr8", "Itpkc__chr7", "Jun__chr4", 
    "Junb__chr8", "Jund__chr8", "Klf2__chr8", "Klf4__chr4", "Klf6__chr13", "Klf9__chr19", 
    "Litaf__chr16", "Lmna__chr3", "Maff__chr15", "Mafk__chr5", "Mcl1__chr3", "Midn__chr10", 
    "Mir22hg__chr11", "Mt1__chr8", "Mt2__chr8", "Myadm__chr7", "Myc__chr15", "Myd88__chr9", 
    "Nckap5l__chr15", "Ncoa7__chr10", "Nfkbia__chr12", "Nfkbiz__chr16", "Nop58__chr1", 
    "Nppc__chr1", "Nr4a1__chr15", "Odc1__chr12", "Osgin1__chr8", "Oxnad1__chr14", 
    "Pcf11__chr7", "Pde4b__chr4", "Per1__chr11", "Phlda1__chr10", "Pnp__chr14", "Pnrc1__chr4", 
    "Ppp1cc__chr5", "Ppp1r15a__chr7", "Pxdc1__chr13", "Rap1b__chr10", "Rassf1__chr9", 
    "Rhob__chr12", "Rhoh__chr5", "Ripk1__chr13", "Sat1__chrX", "Sbno2__chr10", "Sdc4__chr2", 
    "Serpine1__chr5", "Skil__chr3", "Slc10a6__chr5", "Slc38a2__chr15", "Slc41a1__chr1", 
    "Socs3__chr11", "Sqstm1__chr11", "Srf__chr17", "Srsf5__chr12", "Srsf7__chr17", 
    "Stat3__chr11", "Tagln2__chr1", "Tiparp__chr3", "Tnfaip3__chr10", "Tnfaip6__chr2", 
    "Tpm3__chr3", "Tppp3__chr8", "Tra2a__chr6", "Tra2b__chr16", "Trib1__chr15", "Tubb4b__chr2", 
    "Tubb6__chr18", "Ubc__chr5", "Usp2__chr9", "Wac__chr18", "Zc3h12a__chr4", "Zfand5__chr19", 
    "Zfp36__chr7", "Zfp36l1__chr12", "Zfp36l2__chr17", "Zyx__chr6", "Gadd45g__chr13", 
    "Hspe1__chr1", "Ier5__chr1", "Kcne4__chr1")

genes <- sapply(genesChrom, function(x) {
    toupper(strsplit(x, "__")[[1]][1])
})

Data <- as.data.frame(as.matrix(GetAssayData(object = object)))
write.table(genes, paste0(outputDir, "/mouseDissocGenes.tsv"), sep = "\t", quote = FALSE, 
    row.names = FALSE)

## Remove mouse only genes and put the corresponding human
genes <- genes[!genes %in% c("CCRN4L", "MT1", "MT2")]
genes <- c(genes, "NOCT", "MT1A", "MT2A")
cat("Genes from mouse we miss in human:\n")
unname(genes[!genes %in% row.names(Data)])
## Calculate the percentage of UMIs maping on dissociation genes
totalSum <- Matrix::colSums(GetAssayData(object = object))
selection <- Data[genes, ]
selection[is.na(selection)] <- 0
dissociationSums <- colSums(selection)
countSums <- merge(totalSum, dissociationSums, by = "row.names", all = TRUE, sort = FALSE)
rownames(countSums) <- countSums$Row.names
countSums <- countSums[-1]
colnames(countSums) <- c("totalCount", "dissociationCounts")
countSums$percentage <- countSums$dissociationCounts/countSums$totalCount
## Save in meta.data of object
object[["percent.dissoc"]] <- countSums$percentage

## Draw histogram for all samples
percentages <- object$percent.dissoc
hist(percentages, breaks = 100, col = "lightgrey", main = paste("Expression dissociation-affected genes"), 
    xlab = "Ratio of dissociation-affected genes to total gene count", ylab = "Number of cells", 
    xlim = c(0, 0.2))

## Remove object to free up mem
rm(Data)
```

## Keep cells with dissociation percentages below the threshold of 6%
```{r}
object <- subset(x = object, subset = percent.dissoc < 0.06)
object
```

## Variable genes, scaling
```{r}
object <- FindVariableFeatures(object, 
                               selection.method = "vst", 
                               nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(object), 10)
plot1 <- VariableFeaturePlot(object)
LabelPoints(plot = plot1, points = top10, repel = TRUE)

all.genes <- rownames(object)
object <- ScaleData(object, features = all.genes)
```

## Perform linear dimensional reduction
Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset.
```{r}
object <- RunPCA(object, 
                 features = VariableFeatures(object = object))
print(object[["pca"]], 
      dims = 1:5, 
      nfeatures = 8)
VizDimLoadings(object, 
               dims = 1:2, 
               reduction = "pca")
DimPlot(object, reduction = "pca")
DimHeatmap(object, dims = 1, cells = 500, balanced = TRUE)
```

## Elbow plot
```{r}
ElbowPlot(object)
```

## Cluster cells
```{r}
object <- FindNeighbors(object, dims = 1:13)
object <- FindClusters(object, resolution = 0.3)
```

## Run non-linear dimensional reduction UMAP
```{r}
object <- RunUMAP(object, 
                  dims = 1:13)
p <- DimPlot(object, 
             reduction = "umap")
ggplotly(p)
saveRDS(object, file = "sample_b.rds")
```

# Demultiplex cells based on HTO enrichment
* Use the Seurat function HTODemux() to assign single cells back to their sample origins.
```{r}
object <- HTODemux(object, assay = "HTO", positive.quantile = 0.99)
table(object$HTO_classification.global)
table(object$HTO_classification)
```

## Visualize the results on UMAP
```{r}
p <- DimPlot(object,
             reduction = 'umap',
             group.by = "HTO_classification")
ggplotly(p)
```

## Male gene expression
* RPS4Y1 is a gene expressed in males. Not expeced to be expressed in female mesonephros cells
```{r}
FeaturePlot(object, 
            features = "RPS4Y1")
```

## Read the cell ID to subject mapping, as calculated by SCSM
```{r}
cell_id2calledS <- read.table(file = "/exports/sasc/project-371-10X-mesonephros/analysis/SCSM/sample_b_cell_id2calledS.tsv",
                       header = TRUE,
                       sep = "\t")
calledS <- cell_id2calledS$calledS
names(calledS) <- cell_id2calledS$cell_bc
object <- AddMetaData(object = object,
                      metadata = calledS,
                      col.name = "scsm")
p <- DimPlot(object,
             reduction = 'umap',
             group.by = "scsm")
ggplotly(p)
```

## Assess cell Asignment
```{r}
meta.data <- object@meta.data
subjects <- unique(meta.data$scsm)
for (subject in subjects){
  print(subject)
  print(meta.data %>% filter(scsm == subject) %>% count(HTO_classification))
}


saveRDS(object, file = "sample_b.rds")
```

## Separate the mesonephros samples
* SID02: keep cells that are SID02, while negative. Cells are named TK1
* SID03: keep cells that are SID03, while Tag-13-TotalA or negative. Cells are named TR5
* SID01: keep cells that are SID01, while Tag-14-TotalA or negative. Cells are named TN1
```{r}
meta.data <- object@meta.data
object[["cleaned_subject_ids"]] <- case_when(
                                      meta.data$scsm == "SID02" & meta.data$HTO_classification == "Negative" ~ "TK1",
                                      meta.data$scsm == "SID03" & meta.data$HTO_classification %in% c("Tag-13-TotalA", "Negative") ~ "TR5",
                                      meta.data$scsm == "SID01" & meta.data$HTO_classification %in% c("Tag-14-TotalA", "Negative") ~ "TN1"
)

p <- DimPlot(object,
             reduction = 'umap',
             group.by = "cleaned_subject_ids")
ggplotly(p)
saveRDS(object, file = "mesonephros_male.rds")

## Save only the annotated cells
object_meso <- subset(object, cleaned_subject_ids %in% c("TK1", "TR5", "TN1"))
saveRDS(object_meso, file = "meso_b.rds")
```

## Subject counts
* NAs are cells where no subject could be assigned
```{r}
object
table(object$cleaned_subject_ids, useNA='always')
```

### Session Info
```{r}
sessionInfo()
```
